src/golang/lib/workflow/utils/database.go

```diff
@@ -395,35 +362,102 @@ func CreateWorkflowDagResult(
 	)
 }
 
-func UpdateWorkflowDagResultMetadata(
+// UpdateDAGResultMetadata updates the status and execution state of the
+// specified DAGResult. It also creates the relevant notification(s).
+func UpdateDAGResultMetadata(
 	ctx context.Context,
-	workflowDagResultId uuid.UUID,
+	dagResultID uuid.UUID,
 	execState *shared.ExecutionState,
-	workflowDagResultWriter workflow_dag_result.Writer,
+	dagResultRepo repos.DAGResult,
 	workflowReader workflow.Reader,
 	notificationWriter notification.Writer,
-	userReader user.Reader,
-	db database.Database,
-) {
+	DB database.Database,
+) error {
+	txn, err := DB.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer database.TxnRollbackIgnoreErr(ctx, txn)
+
 	changes := map[string]interface{}{
-		workflow_dag_result.StatusColumn:    execState.Status,
-		workflow_dag_result.ExecStateColumn: execState,
+		models.DAGResultStatus:    execState.Status,
+		models.DAGResultExecState: execState,
 	}
 
-	_, err := workflowDagResultWriter.UpdateWorkflowDagResult(
+	dagResult, err := dagResultRepo.Update(
 		ctx,
-		workflowDagResultId,
+		dagResultID,
 		changes,
-		workflowReader,
+		txn,
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := createDAGResultNotification(
+		ctx,
+		dagResult,
 		notificationWriter,
-		userReader,
-		db,
+		workflowReader,
+		txn,
+	); err != nil {
+		return err
+	}
+
+	return txn.Commit(ctx)
+}
+
+func createDAGResultNotification(
+	ctx context.Context,
+	dagResult *models.DAGResult,
+	notificationWriter notification.Writer,
+	workflowReader workflow.Reader,
+	DB database.Database,
+) error {
+	status := dagResult.Status
+	if status != mdl_shared.SucceededExecutionStatus &&
+		status != mdl_shared.FailedExecutionStatus {
+		// Do not create notifications for DAGResults still in progress
+		return nil
+	}
+
+	workflow, err := workflowReader.GetWorkflowByWorkflowDagId(
+		ctx,
+		dagResult.DagID,
+		DB,
 	)
 	if err != nil {
-		log.WithFields(
-			log.Fields{
-				"changes": changes,
-			},
-		).Errorf("Unable to update workflow dag result metadata: %v", err)
+		return err
+	}
+
+	notificationLevel := notification.SuccessLevel
+	notificationContent := fmt.Sprintf(
+		"Workflow %s has succeeded!",
+		workflow.Name,
+	)
+	if status == mdl_shared.FailedExecutionStatus {
+		notificationLevel = notification.ErrorLevel
+		notificationContent = fmt.Sprintf(
+			"Workflow %s has failed.",
+			workflow.Name,
+		)
+	}
+
+	notificationAssociation := notification.NotificationAssociation{
+		Object: notification.WorkflowDagResultObject,
+		Id:     dagResult.ID,
 	}
+
+	// TODO: Create notification for all watchers
```

Comment:
Should we create a task in Linear for this TODO?

Reply:
https://linear.app/aqueducthq/issue/ENG-2079/remove-assumption-of-1-watcher-for-each-workflow-when-creating